core/src/test/java/org/apache/calcite/test/RelBuilderTest.java

```diff
@@ -5585,6 +5590,284 @@ private static RelNode buildCorrelateWithJoin(JoinRelType type, RelBuilder build
+    builder.scan("EMP")
+        .filter(
+            builder.call(SqlStdOperatorTable.GREATER_THAN,
+            builder.field("SAL"), builder.literal(1000)));
+    RelNode commonSubexpr = builder.build();
+
+    // Query 1: Count high earners by department
+    RelNode query1 = builder.push(commonSubexpr)
+        .aggregate(builder.groupKey("DEPTNO"), builder.count())
+        .build();
+
+    // Query 2: Get names of high earners
+    RelNode query2 = builder.push(commonSubexpr)
+        .project(builder.field("ENAME"), builder.field("SAL"))
+        .build();
+
+    // Combine both queries - planner can recognize shared subexpression
+    RelNode combined = builder.combine(query1, query2).build();
+
+    assertThat(combined, instanceOf(Combine.class));
+    assertThat(combined.getInputs(), hasSize(2));
+    // Both queries share the same filter as their base
+    assertThat(combined.getInputs().get(0).getInput(0), instanceOf(Filter.class));
+    assertThat(combined.getInputs().get(1).getInput(0), instanceOf(Filter.class));
+  }
+
+  @Test void testCombineDifferentRowTypes() {
+    // Test that Combine doesn't require inputs to have the same row type (unlike Union)
+    final RelBuilder builder = RelBuilder.create(config().build());
+
+    // Query 1: Returns (ENAME: VARCHAR)
+    RelNode query1 = builder.scan("EMP")
+        .project(builder.field("ENAME"))
+        .build();
+
+    // Query 2: Returns (DEPTNO: INTEGER, AVG_SAL: DECIMAL)
+    RelNode query2 = builder.scan("EMP")
+        .aggregate(builder.groupKey("DEPTNO"),
+            builder.avg(false, "AVG_SAL", builder.field("SAL")))
+        .build();
+
+    // Query 3: Returns just count (COUNT: BIGINT)
+    RelNode query3 = builder.scan("DEPT")
+        .aggregate(builder.groupKey(), builder.count())
+        .build();
+
+    // Combine can handle different row types
+    RelNode combined = builder.combine(query1, query2, query3).build();
+
+    assertThat(combined, instanceOf(Combine.class));
+    assertThat(combined.getInputs(), hasSize(3));
+    // Verify each has different row type
+    assertThat(combined.getInputs().get(0).getRowType().getFieldCount(), is(1));
+    assertThat(combined.getInputs().get(1).getRowType().getFieldCount(), is(2));
+    assertThat(combined.getInputs().get(2).getRowType().getFieldCount(), is(1));
+  }
+
+  @Test void testCombineMetadata() {
+    // Test metadata methods for Combine operator
+    final RelBuilder builder = RelBuilder.create(config().build());
+
+    // Create queries with known characteristics
+    RelNode query1 = builder.scan("EMP")
+        .filter(builder.equals(builder.field("DEPTNO"), builder.literal(10)))
+        .build();
+
+    RelNode query2 = builder.scan("DEPT")
+        .project(builder.field("DNAME"))
+        .build();
+
+    RelNode combined = builder.combine(query1, query2).build();
+
+    // Test that Combine has proper metadata
+    assertThat(combined, instanceOf(Combine.class));
+
+    // Get metadata query
```

Comment:
Nit: I find your comments great, but this one in particular doesn't add much and could be removed IMO